### PR TITLE
Hash search memcache key to limit key length

### LINKED ISSUE #517: Intermittent tracebacks on some searches
Production cnx-archive has been experiencing intermittent tracebacks, some involving searches. The do not seem to be repeatable: searches from the logs that returned a 503, work when tested  -
Example URLs:
```
128.42.169.27 (107.23.217.32, 128.42.169.27) - - [20/Aug/2017:07:47:03 -0500] "GET http://archive.cnx.org/search?q=title:%22Probabil%20and%20Statistics%20for%20engineering%20and%20sciences%22%20&page=1 HTTP/1.1" 200 503 "-" "Mozilla/5.0 (Linux; U; Android 2.2; de-de; HTC Desire HD 1.18.161.2 Build/FRF91) AppleWebKit/533.1 (KHTML, like Gecko) Version/4.0 Mobile Safari/533.1"
128.42.169.27 (107.23.217.32, 128.42.169.27) - - [20/Aug/2017:07:47:03 -0500] "GET http://archive.cnx.org/search?q=title:%22Probabil%20and%20Statistics%20for%20engineering%20and%20sciences%22%20&page=3 HTTP/1.1" 200 503 "-" "Mozilla/5.0 (Linux; U; Android 2.2; de-de; HTC Desire HD 1.18.161.2 Build/FRF91) AppleWebKit/533.1 (KHTML, like Gecko) Version/4.0 Mobile Safari/533.1"
128.42.169.27 (107.23.217.32, 128.42.169.27) - - [20/Aug/2017:07:47:03 -0500] "GET http://archive.cnx.org/search?q=title:%22Probabil%20and%20Statistics%20for%20engineering%20and%20sciences%22%20&page=2 HTTP/1.1" 200 503 "-" "Mozilla/5.0 (Linux; U; Android 2.2; de-de; HTC Desire HD 1.18.161.2 Build/FRF91) AppleWebKit/533.1 (KHTML, like Gecko) Version/4.0 Mobile Safari/533.1"
128.42.169.27 (37.37.34.210, 128.42.169.27) - - [20/Aug/2017:21:30:20 -0500] "GET http://archive.cnx.org/search?q=non-terminating%20rational HTTP/1.1" 200 503 "https://cnx.org/search?q=non-terminating%20rational" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.101 Safari/537.36"
128.42.169.27 (65.96.34.156, 128.42.169.27) - - [20/Aug/2017:22:31:25 -0500] "GET http://archive.cnx.org/search/185cbf87-c72e-48f5-b51e-f14f21b5eabd@10.106:ea5ef43f-dbba-447e-b163-d9dc09b49990@5?q=Basal%20transcription HTTP/1.1" 503 286 "http://cnx.org/contents/GFy_h8cu@10.106:6l70P9u6/Eukaryotic-Transcription" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36 Edge/15.15063"
128.42.169.27 (207.62.146.7, 128.42.169.27) - - [20/Aug/2017:23:42:52 -0500] "GET http://archive.cnx.org/search?q=978-0134112831 HTTP/1.1" 503 286 "https://cnx.org/search?q=978-0134112831" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/603.3.8 (KHTML, like Gecko) Version/10.1.2 Safari/603.3.8"
128.42.169.27 (71.178.15.182, 128.42.169.27) - - [21/Aug/2017:01:25:30 -0500] "GET http://archive.cnx.org/search/14fb4ad7-39a1-4eee-ab6e-3ef2482e3e22@8.108:ada35081-9ec4-4eb8-98b2-3ce350d5427f@3?q=chapter%2018 HTTP/1.1" 503 286 "http://cnx.org/contents/FPtK1zmh@8.108:raNQgZ7E@3/Overview-of-Anatomy-and-Physio" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; rv:47.0) Gecko/20100101 Firefox/47.0"
```

One traceback (may not be any of the above searches):
```
ERROR:waitress:Exception when serving /search
Traceback (most recent call last):
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/waitress/channel.py", line 338, in service
    task.service()
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/waitress/task.py", line 169, in service
    self.execute()
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/waitress/task.py", line 399, in execute
    app_iter = self.channel.server.application(env, start_response)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/paste/deploy/config.py", line 291, in __call__
    return self.app(environ, start_response)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/router.py", line 223, in __call__
    response = self.invoke_subrequest(request, use_tweens=True)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/router.py", line 198, in invoke_subrequest
    response = handle_request(request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/tweens.py", line 20, in excview_tween
    response = handler(request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/router.py", line 145, in handle_request
    view_name
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/view.py", line 541, in _call_view
    response = view_callable(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 327, in attr_view
    return view(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 303, in predicate_wrapper
    return view(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 384, in viewresult_to_response
    result = view(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 506, in _requestonly_view
    response = view(request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/cnxarchive/views.py", line 791, in search
    nocache=params.get('nocache', '').lower() == 'true')
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/cnxarchive/cache.py", line 52, in search
    search_results = mc.get(mc_search_key)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/memcache.py", line 898, in get
    return self._get('get', key)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/memcache.py", line 847, in _get
    self.check_key(key)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/memcache.py", line 1059, in check_key
    % self.server_max_key_length)
MemcachedKeyLengthError: Key length is > 250
```

### PR DESCRIPTION
**Pull request by karenc**
The default memcached server's max key length is 250 characters.  If a
user's search query is really long, archive throws this exception:

```
ERROR:waitress:Exception when serving /search
Traceback (most recent call last):
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/waitress/channel.py", line 338, in service
    task.service()
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/waitress/task.py", line 169, in service
    self.execute()
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/waitress/task.py", line 399, in execute
    app_iter = self.channel.server.application(env, start_response)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/paste/deploy/config.py", line 291, in __call__
    return self.app(environ, start_response)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/router.py", line 223, in __call__
    response = self.invoke_subrequest(request, use_tweens=True)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/router.py", line 198, in invoke_subrequest
    response = handle_request(request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/tweens.py", line 20, in excview_tween
    response = handler(request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/router.py", line 145, in handle_request
    view_name
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/view.py", line 541, in _call_view
    response = view_callable(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 327, in attr_view
    return view(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 303, in predicate_wrapper
    return view(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 384, in viewresult_to_response
    result = view(context, request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/pyramid/config/views.py", line 506, in _requestonly_view
    response = view(request)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/cnxarchive/views.py", line 791, in search
    nocache=params.get('nocache', '').lower() == 'true')
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/cnxarchive/cache.py", line 52, in search
    search_results = mc.get(mc_search_key)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/memcache.py", line 898, in get
    return self._get('get', key)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/memcache.py", line 847, in _get
    self.check_key(key)
  File "/var/cnx/venvs/archive/local/lib/python2.7/site-packages/memcache.py", line 1059, in check_key
    % self.server_max_key_length)
MemcachedKeyLengthError: Key length is > 250
```

To avoid this error, hash the search query using sha1 so the memcache
key stays the same length.

Close #517

### CODE REVIEW

**Comment by pumazi** [comment]
Am I to understand this is an interim fix? A more long term solution might be something that hashes the query and prefixes it as the memcached key?

**Comment by karenc** [comment]
Yes!  We should do that long term.  Given that this PR has not been merged yet, it's not that urgent so we should just do the long term solution.

**Comment by karenc** [comment]
[Fix cnx-archive search error (memcache key too long)](https://trello.com/c/AfWAl5qT/301-fix-cnx-archive-search-error-memcache-key-too-long)

File: cnxarchive/cache.py [hunk 1]
```diff
@@ -69,9 +71,10 @@ def search(query, query_type, nocache=False):
                 # search with one term or one filter, plus query_type
             cache_length = int(settings['search-long-cache-expiration'])
 
-        # store in memcache
-        mc.set(mc_search_key, search_results, time=cache_length,
-               min_compress_len=1024*1024)  # compress when > 1MB
+        if len(mc_search_key) <= mc.server_max_key_length:
+            # store in memcache
+            mc.set(mc_search_key, search_results, time=cache_length,
+                   min_compress_len=1024*1024)  # compress when > 1MB
```
Comment (pumazi) [inline cnxarchive/cache.py]: Is there still the chance that the `mc_search_key` is longer than the max length? I thought we were removing this kind of condition with the use of a hash?

Comment (karenc) [reply cnxarchive/cache.py]: Oh there were 2 places that have this condition... I forgot to remove this one.  I'll update.

**Pull request merged**